src/common/drain/drain.js

```diff
@@ -6,33 +6,66 @@
+  if (!registry[agentIdentifier].get(group)) registry[agentIdentifier].set(group, item)
+}
 
-  var bufferedEventsInGroup = baseEE.backlog[group]
-  var groupHandlers = handlers[group]
-  if (groupHandlers) {
-    // don't cache length, buffer can grow while processing
-    for (var i = 0; bufferedEventsInGroup && i < bufferedEventsInGroup.length; ++i) { // eslint-disable-line no-unmodified-loop-condition
-      emitEvent(bufferedEventsInGroup[i], groupHandlers)
-    }
+function curateRegistry(agentIdentifier) {
+  if (!agentIdentifier) return
+  if (!registry[agentIdentifier]) registry[agentIdentifier] = new Map()
+}
 
-    mapOwn(groupHandlers, function (eventType, handlerRegistrationList) {
-      mapOwn(handlerRegistrationList, function (i, registration) {
-        // registration is an array of: [targetEE, eventHandler]
-        registration[0].on(eventType, registration[1])
-      })
+/**
+ * Drain buffered events out of the EE. Each feature **should** have its events bucket into a "group".  The feature should drain this group explicitly
+ * @param {string} agentIdentifier 
+ * @param {string} group 
+ */
+export function drain(agentIdentifier = "", featureName = 'feature') {
+  curateRegistry(agentIdentifier)
+  // if its not in the registry, that means the instrument file was bypassed.  This could happen in tests, or loaders that directly import the agg
+  if (!agentIdentifier || !registry[agentIdentifier].get(featureName)) return drainGroup(featureName)
+  registry[agentIdentifier].get(featureName).staged = true
+
+  const items = Array.from(registry[agentIdentifier])
+  if (items.every(([key, values]) => values.staged)) {
+    items.sort((a, b) => a[1].priority - b[1].priority)
+    items.forEach(([group]) => {
+      drainGroup(group)
     })
   }
 
-  delete handlers[group]
-  // Keep the group as a property so we know it was created and drained
-  baseEE.backlog[group] = null
+  function drainGroup(group) {
+    const baseEE = agentIdentifier ? ee.get(agentIdentifier) : ee
+    const handlers = defaultRegister.handlers
+    if (!baseEE.backlog || !handlers) return
+
+    var bufferedEventsInGroup = baseEE.backlog[group]
+    var groupHandlers = handlers[group]
+    if (groupHandlers) {
+      // don't cache length, buffer can grow while processing
+      for (var i = 0; bufferedEventsInGroup && i < bufferedEventsInGroup.length; ++i) { // eslint-disable-line no-unmodified-loop-condition
+        emitEvent(bufferedEventsInGroup[i], groupHandlers)
+      }
+
+      mapOwn(groupHandlers, function (eventType, handlerRegistrationList) {
+        mapOwn(handlerRegistrationList, function (i, registration) {
+          // registration is an array of: [targetEE, eventHandler]
+          registration[0].on(eventType, registration[1])
+        })
+      })
+    }
+
+    delete handlers[group]
+    // Keep the group as a property so we know it was created and drained
+    baseEE.backlog[group] = null
+    baseEE.emit('drain-' + group, [])
```

Comment:
One nice thing here is that we can use an event to ensure that features do not start trying to harvest before they are appropriately drained.  (nearly) All features now subscribe to this drain-${name} event before starting their harvest cycle to prevent race conditions starting a harvest without drain having been called.

Reply:
Were features historically gated from harvesting until after RUM because of the other flags? Is this an old system? Are we able to just drain & harvest when features (and their dependecies) are ready, aside from session trace?